In [1]:
from google.colab import files
uploaded = files.upload()

Saving heart_disease_combined.csv to heart_disease_combined.csv


In [2]:
#importing required libraries for data pre-processing.

import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

#importing the required libraries to perform the ML regression for Serum cholestoral.
from sklearn.neural_network import MLPRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.preprocessing import StandardScaler

In [3]:
#loading the final dataset into the dataframe.

heart_data = pd.read_csv("heart_disease_combined.csv")
display(heart_data.head(20))

age  gender  chest_pain  resting_BP   serum_cholestoral  \
0   63.0     1.0         1.0        145.0              233.0   
1   67.0     1.0         4.0        160.0              286.0   
2   67.0     1.0         4.0        120.0              229.0   
3   37.0     1.0         3.0        130.0              250.0   
4   41.0     0.0         2.0        130.0              204.0   
5   56.0     1.0         2.0        120.0              236.0   
6   62.0     0.0         4.0        140.0              268.0   
7   57.0     0.0         4.0        120.0              354.0   
8   63.0     1.0         4.0        130.0              254.0   
9   53.0     1.0         4.0        140.0              203.0   
10  57.0     1.0         4.0        140.0              192.0   
11  56.0     0.0         2.0        140.0              294.0   
12  56.0     1.0         3.0        130.0              256.0   
13  44.0     1.0         2.0        120.0              263.0   
14  52.0     1.0         3.0        172.0              199.0   
15  57.0     1.0         3.0        150.0              168.0   
16  48.0     1.0         2.0        110.0              229.0   
17  54.0     1.0         4.0        140.0              239.0   
18  48.0     0.0         3.0        130.0              275.0   
19  49.0     1.0         2.0        130.0              266.0   

    fasting_blood_sugar  resting_ecg  max_heart_rate  exercise_angina  \
0                   1.0          2.0           150.0              0.0   
1                   0.0          2.0           108.0              1.0   
2                   0.0          2.0           129.0              1.0   
3                   0.0          0.0           187.0              0.0   
4                   0.0          2.0           172.0              0.0   
5                   0.0          0.0           178.0              0.0   
6                   0.0          2.0           160.0              0.0   
7                   0.0          0.0           163.0              1.0   
8                   0.0          2.0           147.0              0.0   
9                   1.0          2.0           155.0              1.0   
10                  0.0          0.0           148.0              0.0   
11                  0.0          2.0           153.0              0.0   
12                  1.0          2.0           142.0              1.0   
13                  0.0          0.0           173.0              0.0   
14                  1.0          0.0           162.0              0.0   
15                  0.0          0.0           174.0              0.0   
16                  0.0          0.0           168.0              0.0   
17                  0.0          0.0           160.0              0.0   
18                  0.0          0.0           139.0              0.0   
19                  0.0          0.0           171.0              0.0   

    oldpeak  slope  num_major_vessels  thal  target  
0       2.3    3.0                0.0   6.0       0  
1       1.5    2.0                3.0   3.0       2  
2       2.6    2.0                2.0   7.0       1  
3       3.5    3.0                0.0   3.0       0  
4       1.4    1.0                0.0   3.0       0  
5       0.8    1.0                0.0   3.0       0  
6       3.6    3.0                2.0   3.0       3  
7       0.6    1.0                0.0   3.0       0  
8       1.4    2.0                1.0   7.0       2  
9       3.1    3.0                0.0   7.0       1  
10      0.4    2.0                0.0   6.0       0  
11      1.3    2.0                0.0   3.0       0  
12      0.6    2.0                1.0   6.0       2  
13      0.0    1.0                0.0   7.0       0  
14      0.5    1.0                0.0   7.0       0  
15      1.6    1.0                0.0   3.0       0  
16      1.0    3.0                0.0   7.0       1  
17      1.2    1.0                0.0   3.0       0  
18      0.2    1.0                0.0   3.0       0  
19      0.6    1.0                0.0   3.0   

In [4]:
#checking the info of the final dataset.

heart_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 720 entries, 0 to 719
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   age                  720 non-null    float64
 1   gender               720 non-null    float64
 2   chest_pain           720 non-null    float64
 3   resting_BP           717 non-null    float64
 4   serum_cholestoral    697 non-null    float64
 5   fasting_blood_sugar  637 non-null    float64
 6   resting_ecg          718 non-null    float64
 7   max_heart_rate       718 non-null    float64
 8   exercise_angina      718 non-null    float64
 9   oldpeak              714 non-null    float64
 10  slope                513 non-null    float64
 11  num_major_vessels    307 non-null    float64
 12  thal                 400 non-null    float64
 13  target               720 non-null    int64  
dtypes: float64(13), int64(1)
memory usage: 78.9 KB


In [5]:
#checking the distribution of the dataset.

heart_data.describe()

age      gender  chest_pain  resting_BP   serum_cholestoral  \
count  720.000000  720.000000  720.000000   717.000000         697.000000   
mean    51.888889    0.738889    3.179167   131.804742         204.774749   
std      9.193720    0.439546    0.953646    18.529331         109.275156   
min     28.000000    0.000000    1.000000    80.000000           0.000000   
25%     45.000000    0.000000    2.000000   120.000000         182.000000   
50%     53.000000    1.000000    4.000000   130.000000         227.000000   
75%     58.000000    1.000000    4.000000   140.000000         270.000000   
max     77.000000    1.000000    4.000000   200.000000         603.000000   

       fasting_blood_sugar  resting_ecg  max_heart_rate  exercise_angina  \
count           637.000000   718.000000      718.000000       718.000000   
mean              0.109890     0.568245      140.565460         0.337047   
std               0.312999     0.833534       25.647172         0.473031   
min               0.000000     0.000000       60.000000         0.000000   
25%               0.000000     0.000000      122.000000         0.000000   
50%               0.000000     0.000000      142.000000         0.000000   
75%               0.000000     1.000000      160.000000         1.000000   
max               1.000000     2.000000      202.000000         1.000000   

          oldpeak       slope  num_major_vessels        thal      target  
count  714.000000  513.000000         307.000000  400.000000  720.000000  
mean     0.789636    1.701754           0.680782    4.985000    0.850000  
std      1.066961    0.585249           0.937101    1.929789    1.076878  
min     -2.600000    1.000000           0.000000    3.000000    0.000000  
25%      0.000000    1.000000           0.000000    3.000000    0.000000  
50%      0.200000    2.000000           0.000000    6.000000    0.500000  
75%      1.500000    2.000000           1.000000    7.000000    1.000000  
max      6.200000    3.000000           3.000000    7.000000    4.000000

In [6]:
#checking for NULL values.

heart_data.isnull().sum()

age                      0
gender                   0
chest_pain               0
resting_BP               3
serum_cholestoral       23
fasting_blood_sugar     83
resting_ecg              2
max_heart_rate           2
exercise_angina          2
oldpeak                  6
slope                  207
num_major_vessels      413
thal                   320
target                   0
dtype: int64

# **Handling NULL values**.


# **1. Resting BP, Resting ECG, Max Heart Rate, Exercise Angina, Old Peak**

In [7]:
#columns to fill NaN values with Mean values.
columns_to_fill = ['resting_BP ', 'serum_cholestoral', 'resting_ecg', 'max_heart_rate','exercise_angina','oldpeak']

#using a for loop to fill NaN values with mean value.
for column in columns_to_fill:
    heart_data[column].fillna(heart_data[column].mean(), inplace=True)

/tmp/ipython-input-7-1331366436.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  heart_data[column].fillna(heart_data[column].mean(), inplace=True)


In [8]:
#re-checking if the NULL values are replaced with Mean values or not.
heart_data.isnull().sum()

age                      0
gender                   0
chest_pain               0
resting_BP               0
serum_cholestoral        0
fasting_blood_sugar     83
resting_ecg              0
max_heart_rate           0
exercise_angina          0
oldpeak                  0
slope                  207
num_major_vessels      413
thal                   320
target                   0
dtype: int64

# **2. SERUM CHOLESTORAL**

In [9]:
#checking for the number of records that have serum_cholestoral as 0.
print(heart_data['serum_cholestoral'].value_counts()[0])

123


***Interpretation***

1.   We got 123 records with 0 value.
2.   Technically the serum cholestoral cannot be 0.
3.   Instead of filling these values with mean or median, we shall use Machine learnign Regression technique to predict these values.
4.   This will allow us to input closer values to these reords.



In [10]:
#splitting the data with serum_cholestoral with/without 0 values.

train_data = heart_data[heart_data['serum_cholestoral'] != 0]
predict_data = heart_data[heart_data['serum_cholestoral'] == 0]

In [11]:
#splitting the train data by considering age, resting_BP and max_heart_rate attributes into consideration.
X = train_data[['age','resting_BP ','max_heart_rate']]
y = train_data['serum_cholestoral']

In [12]:
#splitting the test/train data with 80/20 split.
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

In [13]:
#standardadizing the values.
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_val_scaled = scaler.transform(X_val)

In [14]:
# Train Neural Network (MLPRegressor)
mlp_model = MLPRegressor(hidden_layer_sizes=(64, 32),
                         activation='relu',
                         solver='adam',
                         learning_rate_init=0.001,
                         max_iter=1000,
                         early_stopping=True,
                         random_state=42)

In [15]:
mlp_model.fit(X_train_scaled, y_train)

MLPRegressor(early_stopping=True, hidden_layer_sizes=(64, 32), max_iter=1000,
             random_state=42)

In [16]:
# Predict on validation
y_pred_val = mlp_model.predict(X_val_scaled)

In [17]:
# Evaluation Metrics
rmse = np.sqrt(mean_squared_error(y_val, y_pred_val))
mae = mean_absolute_error(y_val, y_pred_val)
r2 = r2_score(y_val, y_pred_val)

In [18]:
print("\n===== Evaluation Results =====")
print(f"Validation RMSE: {rmse:.2f}")
print(f"Validation MAE : {mae:.2f}")
print(f"Validation R² Score: {r2:.2f}")


===== Evaluation Results =====
Validation RMSE: 55.43
Validation MAE : 42.67
Validation R² Score: -0.12


In [19]:
# Predict missing serumcholestrol values
X_predict_scaled = scaler.transform(predict_data[['age', 'resting_BP ', 'max_heart_rate']])
predicted_cholestrol = mlp_model.predict(X_predict_scaled)


In [20]:
print("\n===== Predicted serumcholestrol values for 0 entries =====")
print(predicted_cholestrol)


===== Predicted serumcholestrol values for 0 entries =====
[296.39732124 243.83169119 256.70800778 268.0863171  238.63730487
 235.75175187 249.49620785 254.07033175 230.92810468 259.85448809
 249.17916332 231.7451354  249.13612535 272.00358965 258.63024494
 226.40787289 232.5364559  228.90637894 237.8246016  256.45769344
 245.56535284 238.70426749 225.90772705 240.85663682 232.96356431
 248.91338082 241.47203284 227.0825921  230.9594346  235.81270796
 290.60693895 261.52559267 234.9418284  231.29218983 334.83201608
 243.87972188 236.79892835 233.940806   249.4350115  316.10666731
 272.66947996 236.76320633 250.82943675 243.40374615 240.9005675
 238.84775263 235.67589847 262.06940378 235.80024061 241.17778242
 277.3407123  233.04386477 239.66436408 275.80744778 280.18191862
 265.53775745 258.7761462  258.33866475 298.0778206  260.33221069
 239.13963449 250.19428096 237.46013067 244.72663031 238.92302009
 239.05890385 249.99699116 265.77317877 293.73082676 241.83560211
 251.10854873 273

In [21]:
# Round the predicted values (remove decimals)
predicted_cholestrol_rounded = np.round(predicted_cholestrol).astype(int)
print(predicted_cholestrol_rounded)

[296 244 257 268 239 236 249 254 231 260 249 232 249 272 259 226 233 229
 238 256 246 239 226 241 233 249 241 227 231 236 291 262 235 231 335 244
 237 234 249 316 273 237 251 243 241 239 236 262 236 241 277 233 240 276
 280 266 259 258 298 260 239 250 237 245 239 239 250 266 294 242 251 273
 269 279 244 240 242 248 245 238 324 270 257 355 262 255 251 243 294 258
 251 259 243 245 244 305 247 313 289 269 248 278 258 313 347 279 258 276
 315 271 269 284 265 258 257 263 258 263 285 285 294 298 283]


In [22]:
# Impute the predicted serumcholestrol values
heart_data.loc[heart_data['serum_cholestoral'] == 0, 'serum_cholestoral'] = predicted_cholestrol_rounded

In [23]:
#checking the distribution of the dataset once we have used regression to impute the serum_cholestoral values.
heart_data.describe()

age      gender  chest_pain  resting_BP   serum_cholestoral  \
count  720.000000  720.000000  720.000000   720.000000         720.000000   
mean    51.888889    0.738889    3.179167   131.804742         249.156693   
std      9.193720    0.439546    0.953646    18.490635          55.173811   
min     28.000000    0.000000    1.000000    80.000000          85.000000   
25%     45.000000    0.000000    2.000000   120.000000         212.750000   
50%     53.000000    1.000000    4.000000   130.000000         244.000000   
75%     58.000000    1.000000    4.000000   140.000000         275.250000   
max     77.000000    1.000000    4.000000   200.000000         603.000000   

       fasting_blood_sugar  resting_ecg  max_heart_rate  exercise_angina  \
count           637.000000   720.000000      720.000000       720.000000   
mean              0.109890     0.568245      140.565460         0.337047   
std               0.312999     0.832373       25.611477         0.472372   
min               0.000000     0.000000       60.000000         0.000000   
25%               0.000000     0.000000      122.000000         0.000000   
50%               0.000000     0.000000      142.000000         0.000000   
75%               0.000000     1.000000      160.000000         1.000000   
max               1.000000     2.000000      202.000000         1.000000   

          oldpeak       slope  num_major_vessels        thal      target  
count  720.000000  513.000000         307.000000  400.000000  720.000000  
mean     0.789636    1.701754           0.680782    4.985000    0.850000  
std      1.062500    0.585249           0.937101    1.929789    1.076878  
min     -2.600000    1.000000           0.000000    3.000000    0.000000  
25%      0.000000    1.000000           0.000000    3.000000    0.000000  
50%      0.200000    2.000000           0.000000    6.000000    0.500000  
75%      1.500000    2.000000           1.000000    7.000000    1.000000  
max      6.200000    3.000000           3.000000    7.000000    4.000000